### DISCUSSION TITLE
Recommendation to perform parallel operations

### DISCUSSION THREAD

**Discussion by Selvam96** [FreeOpcUa/python-opcua#1207] [General]
Hello All,

I'm subscribing to multiple tags. based on particular tag, I'm calling an image analysis function under datachangenotification method During the time it takes to finish the image analysis I Was not able to read and store data of other tags  . I want to read the other tags value as well as to perform image analysis parellely.  Working on closed loop process !. what blocks the operation ?

**Comment by swamper123** [comment]
Hey ho,

running image analysis in the datachange notification is a pretty tough task for a single core. That will be your blocking behaviour.
In an example is written:

>     Do not do expensive, slow or network operation there. Create another
>     thread if you need to do such a thing

In your case I would even say, use another process and try to handle it with that, because you won't have a good performance with threadding as well I would say.

**Reply by Selvam96** [reply]
I'm new to multiprocessing concepts I tried by defining a global list and appended the nod and val to it . when i tried to pass those list as arugs i get error can you post some examples which will be useful for me to know more about it ?

**Reply by AndreasHeine** [reply]
here is a good example series on youtube https://www.youtube.com/watch?v=RR4SoktDQAw

from then its quite simple:
main.py -> creates 2 other processes with a multiprocessing.queue and a multiprocessing.lock
one of them is the opcua client and the other one is your analysis stuff 
each datachange notification appends its data to the queue and the other process takes it from it
the main difference betwen threads and processes in this context each process runs on a different core of your cpu not like thread which run in one python process an one single core
you can also invoke a process pool for your analysis so you can utilize more cores of your cpu

another thing to mention is that in python only processes run really in parallel 

also interesting:
https://www.youtube.com/watch?v=9zinZmE3Ogk

**Reply by Selvam96** [reply]
Thanks for your previous videos it was helpful i understood little bit. dont be harsh I'm still learning .. I dont know how to work with multiprocessing.queue for datachangenotifications method . I tried to do for my function . but i dont know how to pass the queue object to the datachange notification method 

```
class SubHandler():

    def datachange_notification(self, node, val, data):
        print(node,val)
        if val == True and str(node) =="ns=4;s=MAIN.analysisStart"
Previously i called the analysis function here under DCN method but now as suggested trying to use multiprocessing concepts

def analysis1(lock):
        lock.acquire()
       ................................
        lock.release()

v_queue = multiprocessing.queue()
V_process = multiprocessing.process(target=hanler.datachange_notification,args=(v_queue ))
lock =multiprocessing.lock()
A_Process= multiprocessing.process(target=volumeanalysis, args =(lock))
```

**Reply by AndreasHeine** [reply]
last hint:
lock in python dont lock anything! if you use them you can make sure only one process is accessing the Queue
![structure](https://user-images.githubusercontent.com/56362817/104020905-346b1080-51be-11eb-991f-d666766ce822.PNG)

**the rest you should develope on your own because its not part of that library!**

its not ment harsh but @oroulet does not pay well (0$/h) :see_no_evil:

edit:
i should turn off autocorrection in PowerPoint ....

**Reply by Selvam96** [reply]
danke @AndreasHeine . I would like to know how to start the subhandler class using multiprocessing ?

**Reply by oroulet** [reply]
@Selvam96 https://www.google.com/search?q=python+multithreading+tutorial&oq=python+multithreading+tutorial

**Reply by Selvam96** [reply]
```
def subscribe(client,sub_nodes):
    handler = SubHandler()
    sub = client.create_subscription(100, handler)
    sub.subscribe_data_change(sub_nodes)

if __name__ == "__main__":
    
    client = Client("opc.tcp://localhost:4840")
    client.connect()
    analysis_status = client.get_node("ns=4;s=MAIN.analysisStart")
    ntest = client.get_node("ns=4;s=MAIN.ntest")
    sub_nodes = [analysis_status,ntest]
    #p1 = threading.Thread(target=subscribe, args=(client, sub_nodes))
    p1 = multiprocessing.Process(target=subscribe, args=(client, sub_nodes))
    p1.start()
```


I get an error  while working with multiprocess but it works fine with Threading

ie .TypeError: can't pickle _thread.lock objects

**Reply by AndreasHeine** [reply]
thats a miss understanding of threads and processes! you need to learn more about processes and understand the difference to threads... i can only encourage you to think more about program structure even if i send you a code sniped, you need to understand the paradigma behind in the first place!
https://www.youtube.com/watch?v=9zinZmE3Ogk
maybe producer/consumer architecture is a good point to start 
https://stonesoupprogramming.com/2017/09/11/python-multiprocessing-producer-consumer-pattern/comment-page-1/

**Reply by Selvam96** [reply]
Thanks Andreas .. I'm happy to get a such good inputs and support .  I will read more about these concepts

**Comment by oroulet** [comment]
yes put your data in some Queue object and have another thread of proces analyse it
